runtime/src/snapshot_utils.rs

```diff
@@ -1713,6 +1800,106 @@ where
     })
 }
 
+/// Streams snapshot dir files across channel
+/// Follow the flow of streaming_unarchive_snapshot(), but handle the from_dir case.
+fn streaming_snapshot_dir_files<P>(
+    file_sender: Sender<PathBuf>,
+    snapshot_file_path: P,
+    snapshot_version_path: P,
+    account_paths: Vec<PathBuf>,
+) where
+    P: AsRef<Path>,
+{
+    file_sender
+        .send(snapshot_file_path.as_ref().to_path_buf())
+        .unwrap();
+    file_sender
+        .send(snapshot_version_path.as_ref().to_path_buf())
+        .unwrap();
+
+    for account_path in account_paths {
+        for file in fs::read_dir(account_path).unwrap() {
+            let file_path = file.unwrap().path().to_path_buf();
+            file_sender.send(file_path).unwrap();
+        }
+    }
+}
+
+/// Perform the common tasks when deserialize a snapshot.  Handles reading snapshot file, reading the version file,
+/// and then returning those fields plus the rebuilt storage
+fn build_storage_from_snapshot_dir(
+    snapshot_info: &BankSnapshotInfo,
+    measure_name: &'static str,
+    account_paths: &[PathBuf],
+    next_append_vec_id: Arc<AtomicU32>,
+) -> Result<(AccountStorageMap, Measure)> {
+    let bank_snapshot_dir = &snapshot_info.snapshot_dir;
+    let snapshot_file_path = &snapshot_info.snapshot_path();
+    let snapshot_version_path = bank_snapshot_dir.join("version");
+    let (file_sender, file_receiver) = crossbeam_channel::unbounded();
+
+    let accounts_hardlinks = bank_snapshot_dir.join("accounts_hardlinks");
+
+    let account_paths = account_paths.to_vec();
+    let account_paths_set: HashSet<PathBuf> = HashSet::from_iter(account_paths.clone());
+
+    for dir_symlink in fs::read_dir(accounts_hardlinks).unwrap() {
```

Comment:
are these unwraps safe, or should we map err into `SnapshotError::Io`?

Reply:
changed to expect